hadoop-hdfs-project/hadoop-hdfs-rbf/src/main/java/org/apache/hadoop/hdfs/server/federation/router/RouterClientProtocol.java

```diff
@@ -723,6 +723,49 @@ public boolean delete(String src, boolean recursive) throws IOException {
     }
   }
 
+  // Create missing user home dirs for trash paths.
+  // We assume the router is running with super-user privilege (can create
+  // user home dir in /user dir).
+  private void createUserHomeForTrashPath(List<RemoteLocation> locations) throws IOException {
+    List<RemoteLocation> missingUserHomes = new ArrayList<>();
+
+    // Identify missing trash roots
+    for(RemoteLocation loc: locations) {
```

Comment:
I would consider structuring this differently for readability / testability:

1. Map the original list of locations to a List of Home Directories. (i.e. given a list, make another list is easily testable)
2. Filter the list to the list of locations that don't exist. (i.e. given a list)
3. Create those directories that don't exist. (for each element in the list, call function X)

Reply:
Sure, I will split the first for loop into two for loops. The third is already taken care of by invokeSequentialInternal().